src/codegen/util/csv_scanner.cpp

```diff
@@ -0,0 +1,375 @@
+    // in a quoted section, we're done.
+    if (c == '\n' && !in_quote) {
+      break;
+    }
+  }
+
+  // Flush remaining valid bytes
+  if (buffer_pos_ < curr_buffer_pos) {
+    AppendToLineBuffer(buffer_ + buffer_pos_, curr_buffer_pos - buffer_pos_);
+    buffer_pos_ = curr_buffer_pos;
+  }
+
+  // Increment line number
+  line_number_++;
+
+  // If we didn't transfer any bytes to the line buffer, we must have reached an
+  // EOF. If so, return null indicating there are no more lines.
+  if (line_len_ == 0) {
+    return nullptr;
+  }
+
+  // A full line has been transferred to the line buffer, but we also copied the
+  // newline character. Strip it off now.
+  line_len_--;
+  line_[line_len_] = '\0';
+
+  // Done
+  return line_;
+}
+
+void CSVScanner::ProduceCSV(char *line) {
+  const char delimiter = delimiter_;
+  const char quote = quote_;
+  const char escape = escape_;
+
+  // The iterator over characters in the line
+  char *iter = line;
+
+  for (uint32_t col_idx = 0; col_idx < num_cols_; col_idx++) {
+    char *col_begin = iter;
+    char *col_end = nullptr;
+
+    // We need to move col_end to the end of the column's data. Along the way,
+    // we may need to shift data down due to quotes and escapes. Inspired by
+    // Postgres.
+    {
+      char *out = col_begin;
+      while (true) {
+        // This first loop looks for either the delimiter character or the end
+        // of the line, indicating the end of a columns data. It breaks out of
+        // the loop if a quote character is found. It flows into a second loop
+        // whose only purpose is to find the end of the quoted section.
+        while (true) {
+          char c = *iter++;
+
+          // If we see the delimiter character, or the end of the string,
+          // finish
+          if (c == delimiter || c == '\0') {
+            col_end = out;
+            iter--;
+            goto colend;
+          }
+
+          // If we see a quote character, move to the second loop to find the
+          // closing quote.
+          if (c == quote) {
+            break;
+          }
+
+          *out++ = c;
+        }
+
+        while (true) {
+          char c = *iter++;
+
+          // If we see the end of the line *within* a quoted section, throw
+          // error
+          if (c == '\0') {
+            throw Exception(StringUtil::Format(
+                "unterminated CSV quoted field at %u", col_idx));
+          }
+
+          // If we see an escape character within a quoted section, we need to
+          // check if the following character is a quote. If so, we must
+          // escape it
+          if (c == escape) {
+            char next = *iter;
+            if (next == quote || next == escape) {
+              *out++ = next;
+              iter++;
+              continue;
+            }
+          }
+
+          // If we see the closing quote, we're done.
+          if (c == quote) {
+            break;
+          }
+
+          *out++ = c;
```

Comment:
Do I understand correctly, that this line basically writes the current character to the same position in the same buffer - except for the offset that was created by quotes and escapes.

Reply:
Correct, we're writing non-quote characters (unless they're escaped).